CMakeLists.txt

```diff
@@ -13,9 +13,28 @@ endif()
 find_package(ament_cmake REQUIRED)
 find_package(Eigen3 REQUIRED)
 
-install(DIRECTORY include/${PROJECT_NAME}/
-        DESTINATION include/${PROJECT_NAME}
-        FILES_MATCHING PATTERN "*.h")
+add_library(${PROJECT_NAME} INTERFACE
+  include/${PROJECT_NAME}/eigen_stl_containers.h
+  include/${PROJECT_NAME}/eigen_stl_map_container.h
+  include/${PROJECT_NAME}/eigen_stl_vector_container.h
+)
+
+ament_target_dependencies(${PROJECT_NAME} Eigen3)
+ament_export_targets(${PROJECT_NAME}Targets HAS_LIBRARY_TARGET)
+ament_export_dependencies(Eigen3)
+
+install(
+  DIRECTORY include/
+  DESTINATION include
```

Comment:
Since we're modifying this, mind making it install the headers to a unique include directory? It's allows overriding in a merged colcon or ament workspace: https://colcon.readthedocs.io/en/released/user/overriding-packages.html#install-headers-to-a-unique-include-directory

Reply:
Actually, I really do not get the point of why in the world installing into `include/foo/foo/foo.hpp` would be a thing. I have read the paragraph, but it only advises to do so, not why. I have seen (core) packages doing it as well as others not doing it. Also, this is the behavior it had before, did it not?

Reply:
> Actually, I really do not get the point of why in the world installing into include/foo/foo/foo.hpp would be a thing.

It's so that when one package overrides another, the overriding package's headers will be included regardless of the order that include directories are given in CMake. There's more info here: https://colcon.readthedocs.io/en/released/user/overriding-packages.html#include-directory-search-order-problem . ROS 1 solved it by ordering the include directories in `find_package(catking REQUIRED COMPONENTS pkg1 pkg2)`, but that was no longer an option with modern CMake targets.

> I have seen (core) packages doing it as well as others not doing it

Which ones? The core packages should all be using it by now.

Reply:
I have asked this question https://answers.ros.org/question/397356/ros2-conventions-on-installing-headers/